static/app/components/events/eventTags/eventTagsTree.tsx

```diff
@@ -57,136 +86,163 @@ function addToTagTree(
+    const branchRows = getTagTreeRows({
+      ...props,
+      tagKey: t,
+      content: content.subtree[t],
+      spacerCount: spacerCount + 1,
+      isLast: i === subtreeTags.length - 1,
+    });
+    return rows.concat(branchRows);
+  }, []);
+
+  return [
+    <TagTreeRow
+      key={`${tagKey}-${spacerCount}`}
+      tagKey={tagKey}
+      content={content}
+      spacerCount={spacerCount}
+      {...props}
+    />,
+    ...subtreeRows,
+  ];
 }
 
-function createTagTreeItemData(
-  tags: EventTagsTreeProps['tags'],
-  meta: EventTagsTreeProps['meta']
-): [string, TagTreeContent][] {
+/**
+ * Function to create proportional columns for rendering event tags. The columns will not separate
+ * branch tags from their roots, and attempt to be as evenly distributed as possible.
+ * @param {EventTagsTreeProps} props The props for the entire EventTagsTree
+ * @returns {React.ReactNode[]} A list of TreeColumn components to be rendered for this event
+ */
+function assembleTagTreeColumns({
+  meta,
+  tags,
+  ...props
+}: EventTagsTreeProps): React.ReactNode[] {
+  // Create the TagTree data structure using all the given tags
   const tagTree = tags.reduce<TagTree>(
     (tree, tag, i) => addToTagTree(tree, tag, meta?.[i], tag),
     {}
   );
-  return Object.entries(tagTree);
+  // Create a list of TagTreeRow lists, containing every row to be rendered. They are grouped by
+  // root parent so that we do not split up roots/branches when forming columns
+  const tagTreeRowGroups: React.ReactNode[][] = Object.entries(tagTree).map(
+    ([tagKey, content]) => getTagTreeRows({tagKey, content, ...props})
+  );
+  // Get the total number of TagTreeRow components to be rendered, and a goal size for each column
+  const tagTreeRowTotal = tagTreeRowGroups.reduce((sum, group) => sum + group.length, 0);
+  const columnRowGoal = tagTreeRowTotal / COLUMN_COUNT;
+  // Iterate through the row groups, splitting rows into columns when we exceed the goal size
+  const data = tagTreeRowGroups.reduce<TagTreeColumnData>(
+    ({startIndex, runningTotal, columns}, rowList, index) => {
+      runningTotal += rowList.length;
+      // When we reach the goal size (or the last row group), wrap rows in a TreeColumn.
+      if (runningTotal > columnRowGoal || index === tagTreeRowGroups.length - 1) {
+        columns.push(
+          <TreeColumn>{tagTreeRowGroups.slice(startIndex, index)}</TreeColumn>
+        );
+        runningTotal = 0;
+        startIndex = index;
+      }
+      return {startIndex, runningTotal, columns};
+    },
+    {startIndex: 0, runningTotal: 0, columns: []}
+  );
+  return data.columns;
 }
 
-function EventTagsTree({tags, meta, ...props}: EventTagsTreeProps) {
-  const tagTreeItemData = useMemo(() => createTagTreeItemData(tags, meta), [tags, meta]);
+function EventTagsTree(props: EventTagsTreeProps) {
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const tagTreeColumns = useMemo(() => assembleTagTreeColumns(props), []);
```

Comment:
I think assembleTagTreeColumns should be a component since it returns react stuff and do useMemo etc inside of it

Reply:
Good call, made the change 👍